### PR TITLE
expression, parser: simplify the type inferer of builtin funcs

### PR DESCRIPTION
**Pull request by XuHuaiyu**
According to [MySQL source code](https://github.com/mysql/mysql-server/blob/5.7/sql/item.cc#L6326),
the result field type is not alway depending on the types of the arguments.

This commit tries to simplify the type infer process of built-in funcs,
we can _**only**_ focus on `Flen`, `Decimal` and `Flag` in most cases when doing type inferring.

PTAL @shenli @hanfei1991 @zz-jason

### CODE REVIEW

File: expression/builtin_time.go [hunk 1]
```diff
@@ -2023,13 +2023,14 @@ type makeDateFunctionClass struct {
...
-	types.SetBinChsClnFlag(tp)
-	bf, err := newBaseBuiltinFuncWithTp(args, tp, ctx, tpInt, tpInt)
+	bf, err := newBaseBuiltinFuncWithTp(args, tpTime, ctx, tpInt, tpInt)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	tp := bf.tp
+	tp.Tp = mysql.TypeDate
+	tp.Flen = mysql.MaxDateWidth
+	//tp.Decimal = NotFixedDec
```
Comment (hanfei1991) [inline expression/builtin_time.go]: remove this

File: parser/parser.y [hunk 2]
```diff
@@ -4007,12 +4008,24 @@ CastType:
 |	"DATE"
 	{
 		x := types.NewFieldType(mysql.TypeDate)
+		x.Flen = 10
+		x.Decimal = 0
+		x.Charset = charset.CharsetBin
+        x.Collate = charset.CollationBin
```
Comment (hanfei1991) [inline parser/parser.y]: keep alignment

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991

File: expression/builtin_cast.go [hunk 3]
```diff
@@ -170,6 +171,7 @@ type castAsRealFunctionClass struct {
 func (b *castAsRealFunctionClass) getFunction(args []Expression, ctx context.Context) (sig builtinFunc, err error) {
 	bf := baseRealBuiltinFunc{newBaseBuiltinFunc(args, ctx)}
 	bf.tp = b.tp
+	bf.tp.Flen, bf.tp.Decimal = 23, types.UnspecifiedLength
```
Comment (zz-jason) [inline expression/builtin_cast.go]: why specifically set `Flen` and `Decimal` here in `castAsRealFunctionClass` while others like `castAsIntFunctionClass` do not?

`23` may be replaced with `mysql. MaxRealWidth` ?

File: expression/builtin_math.go [hunk 4]
```diff
@@ -849,14 +848,12 @@ type cosFunctionClass struct {
...
-	tp := types.NewFieldType(mysql.TypeDouble)
-	tp.Flen = 23
-	types.SetBinChsClnFlag(tp)
-
-	bf, err := newBaseBuiltinFuncWithTp(args, tp, ctx, tpReal)
+	bf, err := newBaseBuiltinFuncWithTp(args, tpReal, ctx, tpReal)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	bf.tp.Flen = 23
```
Comment (zz-jason) [inline expression/builtin_math.go]: can this magic number be replaced with `mysql.MaxRealWidth`

File: expression/builtin_math.go [hunk 5]
```diff
@@ -1075,13 +1072,12 @@ type tanFunctionClass struct {
...
 func (c *tanFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	tp := types.NewFieldType(mysql.TypeDouble)
-	tp.Flen = 23
-	types.SetBinChsClnFlag(tp)
-	bf, err := newBaseBuiltinFuncWithTp(args, tp, ctx, tpReal)
+	bf, err := newBaseBuiltinFuncWithTp(args, tpReal, ctx, tpReal)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	bf.tp.Flen = 23
```
Comment (zz-jason) [inline expression/builtin_math.go]: ditto

File: expression/builtin_string.go [hunk 6]
```diff
@@ -132,13 +132,11 @@ type lengthFunctionClass struct {
...
 func (c *lengthFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	tp := types.NewFieldType(mysql.TypeLonglong)
-	tp.Flen = 10
-	types.SetBinChsClnFlag(tp)
-	bf, err := newBaseBuiltinFuncWithTp(args, tp, ctx, tpString)
+	bf, err := newBaseBuiltinFuncWithTp(args, tpInt, ctx, tpString)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	bf.tp.Flen = 10
```
Comment (zz-jason) [inline expression/builtin_string.go]: mean `mysql.MaxDateWith`?

File: expression/builtin.go [hunk 7]
```diff
@@ -362,6 +412,21 @@ func (b *baseStringBuiltinFunc) evalDuration(row []types.Datum) (types.Duration,
 	panic("cannot get DURATION result from ClassString expression")
 }
 
+func (b *baseStringBuiltinFunc) getRetTp() *types.FieldType {
+	tp, flen := b.tp, b.tp.Flen
+	if flen >= mysql.MaxBlobWidth {
+		tp.Tp = mysql.TypeLongBlob
+	} else if flen >= 65536 {
```
Comment (XuHuaiyu) [inline expression/builtin.go]: I can't find the real meaning of this number in [MySQL source code](https://github.com/mysql/mysql-server/blob/5.7/sql/item.cc#L6314),
so I didn't convert it to constant.

File: expression/builtin_cast.go (hunk 3 above)
Comment (XuHuaiyu) [inline expression/builtin_cast.go]: yeah, it's wrong behavior.

File: expression/builtin_string.go (hunk 6 above)
Comment (XuHuaiyu) [inline expression/builtin_string.go]: no, `length` returns `int`.

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991 @shenli

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**